doc/tutorials/java-plugins.md

```diff
@@ -11,42 +11,65 @@ If you need a tutorial for using Key and KeySet in Java, please check out the [H
-
-Here we will look into how can we write Java Plugins using the `JNA Binding` executed via the `JNI Plugin`.
+We provide a Java binding for Elektra's API via JNA.
+For more general information about the binding take a look at [this document](java-kdb.md).
 
 ### Writing a Plugin
 
-Under the following [path](/src/bindings/jna/libelektra/src/main/java/org/libelektra) you can find the examples of already existing plugins and you can look into libelektra Java library, which is used for communication with the configuration database.
+To write a plugin, you need to create an implementation of the `org.libelektra.Plugin` interface.
+This is very similar writing any other plugin, except that you use Java instead of C.
+
+The standard API functions of the plugin API all have corresponding methods in the `Plugin` interface:
+
+- `int open(KeySet config, Key errorKey)` is the Java version of [`elektraPluginOpen`](https://doc.libelektra.org/api/latest/html/group__plugin.html#ga23c2eb3584e38a4d494eb8f91e5e3d8d)
+- `int get(KeySet keySet, Key parentKey)` is the Java version of [`elektraPluginGet`](https://doc.libelektra.org/api/latest/html/group__plugin.html#gacb69f3441c6d84241b4362f958fbe313)
+- `int set(KeySet keySet, Key parentKey)` is the Java version of [`elektraPluginSet`](https://doc.libelektra.org/api/latest/html/group__plugin.html#gae65781a1deb34efc79c8cb9d9174842c)
+- `int error(KeySet keySet, Key parentKey)` is the Java version of [`elektraPluginError`](https://doc.libelektra.org/api/latest/html/group__plugin.html#gad74b35f558ac7c3262f6069c5c47dc79)
+- `int close(Key parentKey)` is the Java version of [`elektraPluginClose`](https://doc.libelektra.org/api/latest/html/group__plugin.html#ga1236aefe5b2baf8b7bf636ba5aa9ea29)
+
+Additionally, there is a `@Nonnull String getName()` method.
+This method must return the unique name of the plugin.
+In the C API this would be taken from the contract, but the `process` protocol requires this separately, so we need a separate method for it.
 
-In order to write a new Java Plugin, the new class has to be created under the `plugin` folder. It has also to extend the `Plugin.java` interface, which contains all required methods to communicate with Elektra’s database.
+Otherwise, there are a few differences between implementing a plugin in C and in Java:
 
-The `Plugin.java` interface contains methods for the five basic functions of plugins:
+1. In C it is possible to implement only `elektraPluginGet` and leave the other functions unimplemented.
+   Because of interface inheritance in Java, it is required to implement all 5 method `open`, `get`, `set`, `error` and `close`.
+   Whether a method is actually supported, must be communicated via the contract.
+   Methods that are not supported, should simply return `Plugin.STATUS_SUCCESS`.
```

Comment:
A method that is not marked in the `exports/has` part of the contract shouldn't be called. However, returning `Plugin.STATUS_SUCCESS` is still safer, because if the method is somehow called, simply nothing happens.

Also, if you want to throw something, the appropriate type would be [`UnsupportedOperationException`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/UnsupportedOperationException.html). A `NoSuchMethodError` is a subtype of `Error` and therefore should not be caught and simply lead to a program crash.

Reply:
> because if the method is somehow called, simply nothing happens.

Why would it be "somehow" called? If export/has = 0, we could guarantee that we do not call it? And then it would not matter how the implementation is. A typical implementation in Java would be to throw `NoSuchMethodError`, as calling such a method is basically an assertion error (which also extends `Error`).

The sentences as they are written now sound like we might call it, thus the user must implement it as `Plugin.STATUS_SUCCESS`. This is problematic from maintenance point of view, as someone might think the implementation is as it should be while in fact it is not implemented.

Reply:
> A typical implementation in Java would be to throw NoSuchMethodError, as calling such a method is basically an assertion error (which also extends Error).

Every implementation I've seen would use `UnsupportedOperationException` in this case. AFAIK `NoSuchMethodError` is basically only used by the JVM itself when bytecode for classes is modified at runtime or when using reflection.

---

Thinking about it, the best solution might actually be to use `default` implementations in the interface:

```java
public interface Plugin {
  default int open(KeySet config, Key errorKey) {
    throw new UnsupportedOperationException();
  }
  
  // no default because it must be implemented for the contract
  int get(KeySet keySet, Key parentKey) throws KDBException;

  default int set(KeySet keySet, Key parentKey) {
    throw new UnsupportedOperationException();
  }

  default int error(KeySet keySet, Key parentKey) {
    throw new UnsupportedOperationException();
  }

  default int close(Key parentKey) {
    throw new UnsupportedOperationException();
  }
}
```